### ISSUE TITLE
BI-GRU Embeddings Training

### ISSUE THREAD

**Issue by kinoute** [amaiya/ktrain#30]
While trying to build a binary text classification with a very small dataset (less than 2k files), I was wondering what model will fit best in terms of accuracy of course but also in training time and inference time.

I basically tested everything here, except BERT because I don't have a GPU right now (I can't use cloud services due to confidential datas). I did try distilBERT with another library and get good results.

I will soon I have to make a crucial decision about what model I pick and I struggle for now to decide. Especially because my data can come in many languages. Thats important.

* The LogReg model looks nice but it seems we only have a trainable word embedding with our own vocabulary, which, for my case, with a small dataset, might not been nice (I still have really good results with it). What if suddenly new data in inference time are completely different?

* The NBSVM seems interesting as well, very quick and gives nice results. I just have a False Positive a little high in my opinion (around 14%) and this is the more important metric for me.

* FastText is quick but as I said earlier, I can have text in different languages. There are talks here and there about merging fastText word vectors into a single space vector but I can't use it with this library right now. By default, ktrain seems to detect the language of the dataset. In my case, it can switch from "en" to "fr" depending of the randomly sampled data taken by the function which is problematic too (not sure it downloads the specific French FastText vectors though).

* Finally, the Bidirectional GRU seems to give me good results and low FP (6%). This is similar to what I have with DistilBERT with much longer training and inference time of course.

If I go with BIGRU, I would like to know how it works really inside. I read the paper you linked, I also printed the model layers with `print_layers()`:

```
0 (trainable=False) : <keras.engine.input_layer.InputLayer object at 0x7fe82d337a58>
1 (trainable=True) : <keras.layers.embeddings.Embedding object at 0x7fe82d10b898>
2 (trainable=True) : <keras.layers.core.SpatialDropout1D object at 0x7fe82d0a9828>
3 (trainable=True) : <keras.layers.wrappers.Bidirectional object at 0x7fe82cfa9a58>
4 (trainable=True) : <keras.layers.pooling.GlobalAveragePooling1D object at 0x7fe815ab5ba8>
5 (trainable=True) : <keras.layers.pooling.GlobalMaxPooling1D object at 0x7fe80806e358>
6 (trainable=True) : <keras.layers.merge.Concatenate object at 0x7fe815ab5be0>
7 (trainable=True, wd=None) : <keras.layers.core.Dense object at 0x7fe815ab5cc0>
```

My question is about the embedding layer. It appears that trainable is set to True. Could you confirm me that it means we start with a pre-trained vector from FastText + our own vocabulary taken from the training data (if not already present in FastText), and then during the training time the weights are updated?

This is very important to me because, if I understand correctly, we start with a solid base with a pre-trained vector embedding AND our own vocabulary.

Sorry for the long post.

Thanks in advance for your help!

**Comment by amaiya** [comment]
- **oov**:   If out-of-vocabulary terms are a problem, a good way to handle this is the use of subwords using WordPiece or [BPE](https://github.com/soaxelbrooke/python-bpe).  This is exactly what BERT and DistilBERT do.  When a new never-before-seen word is encountered, it's processed into subwords, which will be in the vocabulary and have a representation.  BERT's vocabulary is only about 30K.

- **embedding layers**:  Note that all embedding layers work the same.  The only difference is the learned vocabulary (which determines the shape of the embedding layer) and how the embedding layer is initialized (either random initialization or initialized with pretrained word vectors).   For non-transfer-learning models, the vocabulary is typically learned from training/validation dataset.  If the training set is small and you want new, never-before-seen words encountered during testing to be represented by a pretrained word vector, you'll need to [fit the tokenizer to a larger corpus than your training set](https://github.com/keras-team/keras/issues/6480).

- **language detection**:  The auto-detected language is used in two ways.  First, it is used to detect non-space-segmented languages like Chinese in which case the text will be  tokenized in a different way.  Second, it is used to select the pretrained BERT model when using `bert`.  If you're not using BERT and you're not worried about Chinese, then the detected language may not matter much.  Also, if you're not happy with the auto-detected language, you can explicitly set the language using the `lang` parameter.

Finally, problems in theory don't always translate to problems in practice.  For some applications, smaller training sets can still capture most of the important vocabulary for the application domain - especially in terms of precision (lower false positives) - in which case `fasttext`, `logreg`, or `nbsvm` might be fine if you're happy with balance of results and speed.
